### ISSUE TITLE
A smart-constructed wrapper around `ArithmeticCircuit` to restrict allowed types of constraints during Haskell compile-time

### ISSUE THREAD

**Issue by vlasin** [zkFold/zkfold-base#252]
In parallel with #249, we must refactor the `acRange` field in the `ArithmeticCircuit` type. Instead of having a map from variables to their upper bound value, it makes sense to have a number of predefined sets (possibly of different dimensions):

$$
 T_1, \dots, T_m
$$

A constraint that we will support has a form

$$
(x_1, \dots, x_{k(j)}) \in T_j, \quad T_j \in \{T_1, \dots, T_m\}.
$$

It makes sense for the sets $T_j$ to parameterise a circuit on a type level, so that adding an unsupported constraint to a circuit will be caught at compile time.

**Comment by TurtlePU** [comment]
If $T_j$ parameterize a circuit on a type level, this should be visible in a `MonadCircuit` API which is accessible via `Symbolic` typeclass, so this leaks into `Symbolic` API. Then this will lead to the same `Sig` problem we had before, this time with lookups which generic context `c` should support.

If this is okay, then sure, why not.

This can be done via `Lookup` marker trait:
```haskell
-- | New definitions
data Lookup = Range Natural | ...
class HasLookup (l :: Lookup) p

-- | Current API updates

class MonadCircuit var a m | m -> a, m -> var where
  -- | Other methods are omitted for clarity
  newRanged :: HasLookup (Range n) m => Witness var a -> m var

class Symbolic c where
  -- | symbolicF is omitted for clarity
  fromCircuit ::
    (forall var a m.
      (MonadCircuit var a m, forall l. HasLookup l c => HasLookup l m) =>
      f var -> m (g var)
    ) -> c f -> c g
  -- ^ Note that this way we have to specify which lookups context `c` `Has`.
  -- This will soon lead to the same problem we had with `Sig`s.

data ArithmeticCircuit (ls :: [Lookup]) a i o = -- ... --
instance In l ls => HasLookup l (ArithmeticCircuit ls a i o)
instance (In l ls, o ~ U1) => HasLookup l (MonadState (ArithmeticCircuit ls a i o))
-- ^ `In l ls` is a type family which compiles iff `l` is in `ls`

instance HasLookup l (Interpreter a)
instance HasLookup l (Witnesses a)
-- ^ lookups are not checked in Interpreter context
```

**Comment by TurtlePU** [comment]
Or maybe it shouldn't be visible in a `MonadCircuit` and we make the type signature permissible (leave it as-is), but we throw an error during circuit compilation if some range constraint is not allowed by the user-specified list of allowed lookups

**Comment by vlasin** [comment]
Hmm, but structured I/O does not leak to Symbolic... I thought it was a similar kind of situation. Currently, the inputs and outputs of the circuit are type parameters. I am thinking about adding the type of allowed constraints as another parameter.

**Comment by TurtlePU** [comment]
So, after discussing this issue a while ago, we came to the conclusion that instead of trying to propagate restrictions inside `Symbolic`, application (and desugaring of) constraints should happen in `compile` function. The best way to do this (keeping the ability to track the types of allowed constraints in Haskell compile-time) is to introduce a sibling type to `ArithmeticCircuit` which would track the constraints and which would be built on post-arithmetization stage.

**Comment by vlasin** [comment]
So, depending on type instantiations in `compile`, we get a different outcome, right? I am wondering, if we want to push this even further. Maybe call `ArithmeticCircuit` an intermediate representation that allows any polynomial, lookup, or fold constraint imaginable, but then we desugar them (or output an error) when we specify the proving system we intend to use. What do you think?

**Comment by TurtlePU** [comment]
Yeah, I think this would be great!

**Comment by vlasin** [comment]
Closing this in favor of #427.
